### PR TITLE
BlockHound used to check the reactivity

### PR DESCRIPTION
**Pull request by kptfh**

### CODE REVIEW

**Comment by Aloren** [comment]
This PR is too big to review. Please split into smaller ones. Thanks.

File: pom.xml [hunk 1]
```diff
@@ -33,6 +33,7 @@
         <joda-time.version>2.10.5</joda-time.version>
         <embedded-aerospike.version>1.33</embedded-aerospike.version>
         <awaitility.version>4.0.1</awaitility.version>
+        <reactor-core.version>3.3.0.RELEASE</reactor-core.version>
```
Comment (Aloren) [inline pom.xml]: you don't need to override dependency. you need just to override the reactor version property from the parent pom

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 2]
```diff
@@ -607,4 +620,57 @@ private Record putAndGetHeader(AerospikeWriteData data, WritePolicy policy) {
 		AerospikePersistentProperty versionProperty = entity.getRequiredVersionProperty();
 		propertyAccessor.setProperty(versionProperty, newRecord.generation);
 	}
+
+	<T> Stream<T> findAllUsingQuery(Class<T> type, Query query) {
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: How are these changes related to blockhound integration?

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java [hunk 3]
```diff
@@ -1,21 +1,13 @@
...
 import com.aerospike.client.policy.GenerationPolicy;
-import org.junit.Before;
 import org.junit.Test;
-import org.springframework.data.aerospike.BaseIntegrationTests;
+import org.springframework.data.aerospike.BaseBlockingIntegrationTests;
 import org.springframework.data.aerospike.SampleClasses.VersionedClass;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-public class AerospikeTemplateDeleteTests extends BaseIntegrationTests {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java]: please extract all refactorings to the separate PRs as it is very difficult to find what is actually related to Blockhound.

File: src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Autowired
+    protected ReactiveAerospikeTemplate reactiveTemplate;
+    @Autowired
+    protected AerospikeReactorClient aerospikeReactorClient;
+
+    @BeforeClass
+    public static void installBlockHound() {
+        log.info("Starting initialization BlockHound libraries to detect blocking call");
+        BlockHound.Builder builder = BlockHound.builder();
```
Comment (bsideup) [inline src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java]: this code is not needed, you can just use `BlockHound.install()`

Also, consider using the JUnit Platform integration (you can still run your JUnit 4 tests with the vintage engine)

Comment (kptfh) [reply src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java]: updated to use BlockHound.install()

Comment (kptfh) [reply src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java]: regarding using of  JUnit Platform integration lets do it in separate pull request.

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java (hunk 3 above)
Comment (kptfh) [reply src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java]: Only this fragment of code is BlockHound related:
`@BeforeClass
    public static void installBlockHound() {
        log.info("Starting initialization BlockHound libraries to detect blocking call");
        BlockHound.install();
        log.info("Finish initialization BlockHound libraries to detect blocking call");
}`
All other code is refactoring that: 
- prepare code to use ReactiveQueryEngine 
- and separate reactive tests from blocking

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 2 above)
Comment (kptfh) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: Only this fragment of code is BlockHound related:

```
@BeforeClass
    public static void installBlockHound() {
        log.info("Starting initialization BlockHound libraries to detect blocking call");
        BlockHound.install();
        log.info("Finish initialization BlockHound libraries to detect blocking call");
    }
```
All other code is refactoring that:
- prepare code to use ReactiveQueryEngine
- and separate reactive tests from blocking

Comment (kptfh) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: let's review it as refactoring

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java (hunk 3 above)
Comment (kptfh) [reply src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateDeleteTests.java]: let's review it as refactoring

File: pom.xml (hunk 1 above)
Comment (kptfh) [reply pom.xml]: fixed

**Comment by Aloren** [comment]
@kptfh all changes related to tests refactoring already merged, you can leave only blockhound related changes here

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 5]
```diff
@@ -607,4 +620,57 @@ private Record putAndGetHeader(AerospikeWriteData data, WritePolicy policy) {
...
+		if ((query.getSort() == null || query.getSort().isUnsorted())
```
Comment (Fetsivalen) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: what if query == null?
as far as I remember it is a usual contract for spring data  "you don't need a query - pass null"

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 6]
```diff
@@ -607,4 +620,57 @@ private Record putAndGetHeader(AerospikeWriteData data, WritePolicy policy) {
...
+
+		Qualifier qualifier = query.getCriteria().getCriteriaObject();
+		Stream<T> results = findAllUsingQuery(type, null, qualifier);
+
+		if (query.getSort() != null && query.getSort().isSorted()) {
+			Comparator comparator = getComparator(query);
+			results = results.sorted(comparator);
+		}
+
+		if(query.hasOffset()) {
```
Comment (Fetsivalen) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: is it ok to use methods like skip and limit here and return stream from the method at the same time?

Since the execution of the whole stream is lazy you may have different results based on collectors based on collectors you apply

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 7]
```diff
@@ -323,4 +331,62 @@ private Throwable translateError(Throwable e) {
         return e;
     }
 
+    <T> Flux<T> findAllUsingQuery(Class<T> type, Query query) {
+        if ((query.getSort() == null || query.getSort().isUnsorted())
```
Comment (Fetsivalen) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: null check?

File: src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    @BeforeClass
+    public static void installBlockHound() {
+        log.info("Starting initialization BlockHound libraries to detect blocking call");
+        BlockHound.install();
+        log.info("Finish initialization BlockHound libraries to detect blocking call");
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void shouldFailAsBlocking(){
```
Comment (Fetsivalen) [inline src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java]: how this test is related to the spring data aerospike repository?
please remove it from here
you test the BlockHound not the spring data aerospike

Comment (bsideup) [reply src/test/java/org/springframework/data/aerospike/BaseReactiveIntegrationTests.java]: This is a,very important test that checks that you installed BlockHound correctly, to avoid false positives
